src/internal_rules.jl

```diff
@@ -761,47 +775,73 @@ function EnzymeRules.augmented_primal(
             end
         end
     end
-    cache = if isa(A, Const)
-        nothing
-    else
-        dfact
-    end
 
+    cache = isa(A, Const) ? nothing : (fact_returned, dfact)
     return EnzymeRules.AugmentedReturn(fact, dfact, cache)
 end
 
 function EnzymeRules.reverse(
     config,
     ::Const{typeof(cholesky)},
     RT::Type,
-    dfact,
+    cache,
     A::Annotation{<:Union{Matrix,LinearAlgebra.RealHermSym{<:Real,<:Matrix}}};
     kwargs...)
-
     if !(RT <: Const) && !isa(A, Const)
+        fact, dfact = cache
         dAs = EnzymeRules.width(config) == 1 ? (A.dval,) : A.dval
         dfacts = EnzymeRules.width(config) == 1 ? (dfact,) : dfact
 
         for (dA, dfact) in zip(dAs, dfacts)
             _dA = dA isa LinearAlgebra.RealHermSym ? dA.data : dA
             if _dA !== dfact.factors
-                _dA .+= dfact.factors
+                Ā = _cholesky_pullback_shared_code(fact, dfact)
+                idx = diagind(Ā)
+                @views Ā[idx] .= real.(Ā[idx]) ./ 2
+                _dA .+= UpperTriangular(Ā)
```

Comment:
I don't think any of this should be here when `A` is a `Matrix`, but this doesn't look right to me for `RealHermSym`. I would expect something like
```suggestion
                _dA .+= LowerTriangular(Ā) + UpperTriangular(Ā)' - real(Diagonal(Ā))
```
(which in ChainRules would be the pullback of the `Symmetric` constructor), but I would move that to `_cholesky_pullback_shared_code`.

Reply:
Do you have a test for this? I'm not sure I understand why the constructor should enter (see also the recent Slack discussion for `sum` [here](https://julialang.slack.com/archives/C01J3R4FHB3/p1709555032435419)).

Reply:
When working out AD rules, we generally treat tangents as dense arrays (i.e. we use the same semantics as ChainRules), and that's what the function `_cholesky_forward` is doing. But after doing that, you need to convert to Enzyme's semantics. That's effectively the ChainRules-style pullback of the constructor (which is what this code does).

Reply:
Making the suggested change breaks a bunch of tests (the original test that was here before this PR, as well as its modified version [here](https://github.com/simsurace/Enzyme.jl/blob/33e48afdd7ea345f69218ba6630d4a36f337e7c7/test/internal_rules.jl#L319), and [this testset](https://github.com/simsurace/Enzyme.jl/blob/33e48afdd7ea345f69218ba6630d4a36f337e7c7/test/internal_rules.jl#L331)). While I'm not completely sure of the former, I'm close to certain that the latter test is correct because it compares differentiating two functions that are mathematically equivalent (both representing `A -> sum(abs2, A * A')`), one of which passes through `cholesky`, and therefore shouldn't break under any circumstance.